### PR TITLE
Type the router package

### PR DESCRIPTION
**Pull request by youknowriad**
I wanted to make some changes to the router package but I noticed that it was not typed yet (which I needed for my changes). so this is just a small PR that updates that package to typescript.

### CODE REVIEW

File: packages/router/src/history.ts [hunk 1]
```diff
@@ -2,12 +2,16 @@
  * External dependencies
  */
 import { createBrowserHistory } from 'history';
+import type { BrowserHistory } from 'history';
```
Comment (sirreal) [inline packages/router/src/history.ts]: **FYI** (not a change request)

Import and type import used to have to be kept separate, but now they _could_ be merged into one and each name can be annotated as a type import:

```suggestion
import { createBrowserHistory, type BrowserHistory } from 'history';
```

File: packages/router/src/history.ts [hunk 2]
```diff
@@ -2,12 +2,16 @@
...
 
 /**
  * WordPress dependencies
  */
 import { buildQueryString } from '@wordpress/url';
 
+export type EnhancedHistory = BrowserHistory & {
+	getLocationWithParams: () => Location;
+};
 const history = createBrowserHistory();
```
Comment (sirreal) [inline packages/router/src/history.ts]: I recommend sticking to interfaces when possible:

```suggestion
export interface EnhancedHistory extends BrowserHistory {
	getLocationWithParams: () => Location;
}

const history = createBrowserHistory();
```

File: packages/router/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,16 @@
+{
+	"$schema": "https://json.schemastore.org/tsconfig.json",
+	"extends": "../../tsconfig.base.json",
+	"compilerOptions": {
+		"rootDir": "src",
+		"declarationDir": "build-types",
+		"types": [ "gutenberg-env" ],
+		"checkJs": false
```
Comment (sirreal) [inline packages/router/tsconfig.json:9]: Don't allow JS in the package:

```suggestion
		"allowJs": false,
		"checkJs": false
```

**Review by sirreal — APPROVED** [review]
Good stuff, thank you.

I left some suggestions, but nothing major in the changes. 2 additional changes I think would be good:

- A changelog note
- I think `"types": "build-types"` should be added to the package.json to expose this package's types like other packages do.

File: packages/router/src/history.ts (hunk 1 above)
Comment (youknowriad) [reply packages/router/src/history.ts]: Let's include the change anyway 👍

File: packages/router/CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+## Enhancements
```
Comment (sirreal) [inline packages/router/CHANGELOG.md]: Wrong level:
```suggestion
### Enhancements
```

**Pull request merged**